File: go/classic/maxEnvelopes_354.go
package classic

import (
	"sort"
)

func maxEnvelopes(envelopes [][]int) int {
	// envelopes[0] 相等 说明宽度相等 这个时候 只需要更长的排在后面即可
	// envelopes[0] 不等 说明宽度不等 这个时候 只需要只需要根据长度大小从大到小排序即可
	sort.Slice(envelopes, func(i, j int) bool {
		if envelopes[i][0] == envelopes[j][0] {
			return envelopes[i][1] > envelopes[j][1]
		} else {
			return envelopes[i][0] < envelopes[j][0]
		}
	})

	// 因为现在这样排序之后 信封的宽度 一定是满足顺序的 那么只需要判断长度 能够形成的最长的递增子序列是多长
	dp := make([]int, len(envelopes))
	// 1 个数字也能有一个长度
	for i := 0; i < len(dp); i++ {
		dp[i] = 1
	}

	res := 0
	for i := 0; i < len(dp); i++ {
		tmp := 0
		for j := 0; j < i; j++ {
			if envelopes[i][1] > envelopes[j][1] {
				tmp = max(tmp, dp[j])
			}
		}
		dp[i] = tmp + 1
		res = max(res, dp[i])
	}

	return res
}

File: java/src/MaxEnvelopes_354.java
import java.lang.reflect.Array;
import java.util.Arrays;

public class MaxEnvelopes_354 {

    public int maxEnvelopes(int[][] envelopes) {
        if (envelopes == null || envelopes.length == 0) {
            return -1;
        }
        Arrays.sort(envelopes, (a, b) -> {
            if (a[0] == b[0]) {
                return b[1] - a[1];
            }
            return a[0] - b[0];
        });

        // 现在就变成了一个寻找最长递增子序列的问题
        // 因为他要求的是前一个的长已经满足 所以是宽度上的最长递增西序列
        int res = 1;
        int[] dp = new int[envelopes.length];
        // 每个数的最长递增子序列肯定是自己 初始化
        Arrays.fill(dp, 1);
        for (int i = 0; i < envelopes.length; i++) {
            for (int j = 0; j < i; j++) {
                if (envelopes[j][1] < envelopes[i][1]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
                res = Math.max(dp[i], res);
            }
        }
        return res;
    }
}
